### PR TITLE
UL-270: Add cursor system, delete Page, test Item interface

### PR DESCRIPTION
**Pull request by juliengroch**

### CODE REVIEW

**Comment by juliengroch** [comment]
ready to review @thoas @gillesfabio

File: stores.go [hunk 1]
```diff
@@ -44,3 +55,20 @@ func (s *GORMStore) Paginate(limit, offset int64, count *int64) error {
...
+
+	q = q.Limit(int(limit))
+
+	if reverse {
+		q = q.Where(fmt.Sprintf("%s < ?", fieldName), cursor)
+	} else {
+		q = q.Where(fmt.Sprintf("%s > ?", fieldName), cursor)
+	}
+
+	q = q.Find(s.items)
```
Comment (gillesfabio) [inline stores.go]: Why not returning GORM error here? `return q.Find(s.items).Error`

File: constants.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package paging
+
+// pagination system type
+const (
```
Comment (gillesfabio) [inline constants.go]: Maybe "better" to name constants like this:

* `TypeOffset`
* `TypeCursor`
* `CursorModeDate`
* `CursorModeID`

IMHO, cleaner naming and much more easier for completion (If I type `Type`, completion will show me all Type constants).

That's purely subjective. My peanuts :)

File: paginator.go [hunk 3]
```diff
@@ -3,177 +3,271 @@ package paging
...
+		return nil, err
+	}
+
+	np.NextURI = p.MakeNextURI()
+
+	return &np, nil
+}
+
+// HasPrevious returns false, previous page is not available on cursor system
+func (p *CursorPaginator) HasPrevious() bool {
```
Comment (gillesfabio) [inline paginator.go]: You don't need pointer receiver here. `func (CursorPaginator) HasPrevious() bool` is enough. The same for `HasNext()` and `MakePreviousURI()`.

File: stores.go [hunk 4]
```diff
@@ -44,3 +55,20 @@ func (s *GORMStore) Paginate(limit, offset int64, count *int64) error {
 
 	return nil
```
Comment (gillesfabio) [inline stores.go]: Maybe we should at least return `q.Error` here (instead of useless `nil`).

File: utils.go [hunk 5]
```diff
@@ -65,12 +66,82 @@ func GetOffsetFromRequest(request *http.Request, options *Options) int64 {
...
+		}
+	} else {
+		cursor = 0
+	}
+
+	return cursor
+}
+
+// GenerateOffsetURI generates the pagination URI.
+func GenerateOffsetURI(limit int64, offset int64, options *Options) string {
```
Comment (gillesfabio) [inline utils.go]: As we don't handle `nil` with `options`, maybe better to pass `options Options` (so as a value) just for safer code.

**Pull request merged**